From 057a1c93db67aeea196913bc6a0657b31982f2f2 Mon Sep 17 00:00:00 2001
From: Christina Jacob <cjacob@marvell.com>
Date: Mon, 24 Jun 2019 15:02:17 +0530
Subject: [PATCH 336/386] octeontx2-pf: Support to display current settings of
 a vf network interface via ethtool

Support ethtool <interface> on both pf's vf interfaces as well as on lbk
interfaces.

Change-Id: I2b5e03d345900ab631466648e533da31d48e6b92
Signed-off-by: Christina Jacob <cjacob@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/12460
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 .../marvell/octeontx2/nic/otx2_ethtool.c      | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 5466b3f6e2ac..084cd3f40473 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -1478,6 +1478,25 @@ static int otx2vf_get_sset_count(struct net_device *netdev, int sset)
 	       otx2_n_queue_stats * (vf->hw.rx_queues + vf->hw.tx_queues);
 }
 
+static int otx2vf_get_link_ksettings(struct net_device *netdev,
+				     struct ethtool_link_ksettings *cmd)
+{
+	struct otx2_nic *pfvf = netdev_priv(netdev);
+
+	if (pfvf->pdev->device ==  PCI_DEVID_OCTEONTX2_RVU_AFVF) {
+		cmd->base.port = PORT_OTHER;
+		if (!netif_running(netdev)) {
+			cmd->base.duplex = DUPLEX_UNKNOWN;
+			cmd->base.speed = SPEED_UNKNOWN;
+		} else {
+			cmd->base.duplex = DUPLEX_FULL;
+			cmd->base.speed = SPEED_100000;
+		}
+	} else {
+		return	otx2_get_link_ksettings(netdev, cmd);
+	}
+	return 0;
+}
 static const struct ethtool_ops otx2vf_ethtool_ops = {
 	.get_drvinfo		= otx2vf_get_drvinfo,
 	.get_strings		= otx2vf_get_strings,
@@ -1497,6 +1516,7 @@ static const struct ethtool_ops otx2vf_ethtool_ops = {
 	.set_coalesce		= otx2_set_coalesce,
 	.get_pauseparam		= otx2_get_pauseparam,
 	.set_pauseparam		= otx2_set_pauseparam,
+	.get_link_ksettings     = otx2vf_get_link_ksettings,
 };
 
 void otx2vf_set_ethtool_ops(struct net_device *netdev)
-- 
2.17.1

